src/jit/objectalloc.cpp

```diff
@@ -16,57 +16,227 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 //             variables that can be allocated on the stack.
-//
-// Assumptions:
-//    Must be run after the dominators have been computed (we need this
-//    information to detect loops).
+
 void ObjectAllocator::DoAnalysis()
 {
     assert(m_IsObjectStackAllocationEnabled);
-    assert(comp->fgDomsComputed);
-    // TODO-ObjectStackAllocation
-    NYI("DoAnalysis");
+    assert(!m_AnalysisDone);
+
+    if (comp->lvaCount > 0)
+    {
+        m_EscapingPointers         = BitVecOps::MakeEmpty(&m_bitVecTraits);
+        m_ConnGraphAdjacencyMatrix = new (comp->getAllocator()) BitSetShortLongRep[comp->lvaCount];
+
+        MarkEscapingVarsAndBuildConnGraph();
+        ComputeEscapingNodes(&m_bitVecTraits, m_EscapingPointers);
+    }
+
+    m_AnalysisDone = true;
+}
+
+//------------------------------------------------------------------------------
+// MarkEscapingVarsAndBuildConnGraph : Walk the trees of the method and mark any ref/byref/i_impl
+//                                     local variables that may escape. Build a connection graph
+//                                     for ref/by_ref/i_impl local variables.
+//
+// Arguments:
+//    sourceLclNum  - Local variable number of the edge source
+//    targetLclNum  - Local variable number of the edge target
+//
+// Notes:
+//     The connection graph has an edge from local variable s to local variable t if s may point
+//     to the objects t points to at some point in the method. It's a simplified version
+//     of the graph described in this paper:
+//     https://www.cc.gatech.edu/~harrold/6340/cs6340_fall2009/Readings/choi99escape.pdf
+//     We currently don't have field edges and the edges we do have are called "deferred" in the paper.
+
+void ObjectAllocator::MarkEscapingVarsAndBuildConnGraph()
+{
+    BuildConnGraphVisitorCallbackData callbackData(this);
+
+    for (unsigned int lclNum = 0; lclNum < comp->lvaCount; ++lclNum)
+    {
+        var_types type = comp->lvaTable[lclNum].TypeGet();
+
+        if (type == TYP_REF || type == TYP_I_IMPL || type == TYP_BYREF)
+        {
+            m_ConnGraphAdjacencyMatrix[lclNum] = BitVecOps::MakeEmpty(&m_bitVecTraits);
+
+            if (comp->lvaTable[lclNum].lvAddrExposed)
+            {
+                JITDUMP("   V%02u is address exposed\n", lclNum);
+                MarkLclVarAsEscaping(lclNum);
+            }
+        }
+        else
+        {
+            // Variable that may not point to objects will not participate in our analysis.
+            m_ConnGraphAdjacencyMatrix[lclNum] = BitVecOps::UninitVal();
+        }
+    }
+
+    BasicBlock* block;
+
+    foreach_block(comp, block)
+    {
+        for (GenTreeStmt* stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
+        {
+            const bool lclVarsOnly  = true;
+            const bool computeStack = true;
+
+            comp->fgWalkTreePre(&stmt->gtStmtExpr, BuildConnGraphVisitor, &callbackData, lclVarsOnly, computeStack);
```

Comment:
I'd recommend using GenTreeVisitor instead of fgWalk functions (which are wrappers around GenTreeVisitor). It tends to be slightly faster and avoids the hassle with extra static functions, casts to walk data and whatnot.

Reply:
Thanks for the suggestion. Done.